File: langs/javascript/src/cli/commands/encrypt.ts
/**
 * `dotlyte encrypt` — Encrypt sensitive values in .env files.
 *
 * Usage:
 *   dotlyte encrypt <file> [--keys KEY1,KEY2,...] [--output <file>]
 *
 * Encrypts values for specified keys (or all if not specified).
 */

import { existsSync, readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { encryptValue, resolveEncryptionKey, generateKey, isEncryptedValue } from "../../encryption.js";
import { isAutoSensitive } from "../../masking.js";

export async function encrypt(args: string[]): Promise<void> {
  if (args.length === 0) {
    throw new Error("Usage: dotlyte encrypt <file> [--keys KEY1,KEY2] [--output <file>]");
  }

  const inputFile = args[0]!;
  const keysArg = getArg(args, "--keys");
  const outputFile = getArg(args, "--output");

  const inputPath = resolve(inputFile);
  if (!existsSync(inputPath)) {
    throw new Error(`File not found: ${inputPath}`);
  }

  // Resolve encryption key
  let encKey = resolveEncryptionKey();
  if (!encKey) {
    console.log("⚠️  No encryption key found. Generating a new one...\n");
    const passphrase = generateKey();
    console.log(`  🔑 New key: ${passphrase}`);
    console.log(`  Set DOTLYTE_KEY=${passphrase} in your environment`);
    console.log(`  Or add it to .dotlyte-keys: default=${passphrase}\n`);

    // Use the generated key for this operation
    const { scryptSync } = await import("node:crypto");
    encKey = scryptSync(passphrase, "dotlyte-v2", 32);
  }

  const content = readFileSync(inputPath, "utf-8");
  const targetKeys = keysArg ? new Set(keysArg.split(",").map((k) => k.trim())) : null;
  const lines = content.split("\n");
  const encryptedLines: string[] = [];
  let encryptedCount = 0;

  console.log(`\n🔒 dotlyte encrypt\n`);
  console.log(`  File: ${inputFile}`);

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) {
      encryptedLines.push(line);
      continue;
    }

    const stripped = trimmed.startsWith("export ") ? trimmed.slice(7).trim() : trimmed;
    const eqIdx = stripped.indexOf("=");
    if (eqIdx <= 0) {
      encryptedLines.push(line);
      continue;
    }

    const key = stripped.slice(0, eqIdx).trim();
    let value = stripped.slice(eqIdx + 1).trim();

    // Remove quotes
    if (value.length >= 2 && value[0] === value[value.length - 1] && (value[0] === '"' || value[0] === "'")) {
      value = value.slice(1, -1);
    }

    // Skip already encrypted
    if (isEncryptedValue(value)) {
      encryptedLines.push(line);
      continue;
    }

    // Determine if this key should be encrypted
    const shouldEncrypt = targetKeys
      ? targetKeys.has(key)
      : isAutoSensitive(key);

    if (shouldEncrypt) {
      const encrypted = encryptValue(value, encKey);
      encryptedLines.push(`${key}=${encrypted}`);
      encryptedCount++;
      console.log(`  🔐 ${key} — encrypted`);
    } else {
      encryptedLines.push(line);
    }
  }

  const outPath = outputFile ? resolve(outputFile) : inputPath.replace(/\.env/, ".env.encrypted");
  writeFileSync(outPath, encryptedLines.join("\n"), "utf-8");

  console.log(`\n  ✅ Encrypted ${encryptedCount} value(s) → ${outPath}\n`);
}

function getArg(args: string[], flag: string): string | undefined {
  const idx = args.indexOf(flag);
  return idx >= 0 && idx + 1 < args.length ? args[idx + 1] : undefined;
}

File: langs/python/src/dotlyte/cli/encrypt.py
"""``dotlyte encrypt`` — Encrypt sensitive values in .env files."""

from __future__ import annotations

import re
import sys
from pathlib import Path

from dotlyte.encryption import (
    encrypt_value,
    generate_key,
    resolve_encryption_key,
    _derive_key,
)
from dotlyte.masking import _SENSITIVE_PATTERNS


def run_encrypt(
    *,
    file: str,
    keys: str | None = None,
    output: str | None = None,
) -> None:
    """Encrypt values for specified keys (or auto-detect sensitive keys).

    Args:
        file: Path to the .env file.
        keys: Comma-separated list of keys to encrypt (optional).
        output: Output file path (optional).

    """
    input_path = Path(file).resolve()
    if not input_path.is_file():
        print(f"❌ File not found: {file}", file=sys.stderr)
        sys.exit(1)

    # Resolve or generate encryption key
    enc_key = resolve_encryption_key(base_dir=str(input_path.parent))
    if enc_key is None:
        print("⚠️  No encryption key found. Generating a new one...\n")
        passphrase = generate_key()
        print(f"  🔑 New key: {passphrase}")
        print(f"  Set DOTLYTE_KEY={passphrase} in your environment")
        print(f"  Or add it to .dotlyte-keys: {passphrase}\n")
        enc_key = _derive_key(passphrase)

    target_keys = {k.strip() for k in keys.split(",")} if keys else None

    content = input_path.read_text(encoding="utf-8")
    lines = content.splitlines()
    encrypted_lines: list[str] = []
    encrypted_count = 0

    print(f"\n🔒 dotlyte encrypt\n")
    print(f"  File: {file}")

    enc_pattern = re.compile(r"^ENC\[")

    for line in lines:
        trimmed = line.strip()
        if not trimmed or trimmed.startswith("#"):
            encrypted_lines.append(line)
            continue

        stripped = trimmed[7:].strip() if trimmed.startswith("export ") else trimmed
        eq_idx = stripped.find("=")
        if eq_idx <= 0:
            encrypted_lines.append(line)
            continue

        key_part = stripped[:eq_idx].strip()
        value_part = stripped[eq_idx + 1 :].strip()

        # Remove quotes
        if (
            len(value_part) >= 2
            and value_part[0] == value_part[-1]
            and value_part[0] in ('"', "'")
        ):
            value_part = value_part[1:-1]

        # Skip already encrypted
        if enc_pattern.match(value_part):
            encrypted_lines.append(line)
            continue

        should_encrypt = (
            key_part in target_keys
            if target_keys
            else _is_auto_sensitive(key_part)
        )

        if should_encrypt:
            encrypted_val = encrypt_value(value_part, enc_key)
            encrypted_lines.append(f"{key_part}={encrypted_val}")
            encrypted_count += 1
            print(f"  🔐 {key_part} — encrypted")
        else:
            encrypted_lines.append(line)

    out_path = (
        Path(output).resolve()
        if output
        else input_path.with_suffix(".encrypted")
    )
    out_path.write_text("\n".join(encrypted_lines) + "\n", encoding="utf-8")

    print(f"\n  ✅ Encrypted {encrypted_count} value(s) → {out_path}\n")


def _is_auto_sensitive(key: str) -> bool:
    """Return True if *key* matches a built-in sensitive pattern."""
    for pattern in _SENSITIVE_PATTERNS:
        if pattern.search(key):
            return True
    return False
